### PR TITLE
feat!:  necessary api changes for React 19 compatibility

### PR DESCRIPTION
**Pull request by phryneas**
This release has to break some old APIs to stay compatible with React 19.

## `render` is now `async` and should be `await`ed

This is the core change - due to the impementation of sibling prerendering on the React side, rendering has become more `async` and tests that interacted with `render` in a synchronous fashion would end up with not-resolving suspense boundaries.

Please adjust your tests accordingly:

```diff
  const {takeRender, render} = createRenderStream({ /* ... */ })
-  const utils = render(<Counter />)
+  const utils = await render(<Counter />)
```

## enforcing of `IS_REACT_ACT_ENVIRONMENT == false`

In combination with [issues we have have seen in the past](https://github.com/facebook/react/issues/29855), we have deduced that the testing approach of this library only really works in a "real-world" scenario, not in an `act` environment.

As a result, we will now throw an error if you try to use it in an environment where `IS_REACT_ACT_ENVIRONMENT` is truthy.

We are shipping a new tool, `disableActEnvironment` to prepare your environment for the duration of a test in a safe manner.

This helper can either be used with explicit resource management using the `using` keyword:

```ts
test('my test', () => {
  using _disabledAct = disableActEnvironment()

  // your test code here

  // as soon as this scope is left, the environment will be cleaned up
})
```

of by manually calling `cleanup`:

```ts
test('my test', () => {
  const {cleanup} = disableActEnvironment()

  try {
    // your test code here
  } finally {
    cleanup()
  }
})
```

This function does not only adjust your `IS_REACT_ACT_ENVIRONMENT` value, but it will also temporarily adjust the `@testing-library/dom` configuration in a way so that e.g. calls to `userEvent` will not automatically be wrapped in `act`.

Of course you can also use this tool on a per-file basis instead of a per-test basis, but keep in mind that doing so will break any React Testing Library tests that might be in the same file.

## `render` is now it's own implementation

Previously, we used the `render` function of React Testing Library, but with the new restrictions this is no longer possible and we are shipping our own implementation.

As a result, some less-common options are not supported in the new implementation. 
If you have a need for these, please let us know!

* `hydrate` was removed
* `legacyRoot` was removed. If you are using React 17, it will automatically switch to `ReactDOM.render`, otherwise we will use `createRoot`

> [!CAUTION]
> React 17 does not look for `IS_REACT_ACT_ENVIRONMENT` to determine if it is running in an `act`-environment, but rather `typeof jest !== "undefined"`.
> If you have to test React 17, we recommend to patch it with a [`patch-package` patch](https://github.com/apollographql/apollo-client/blob/8a4738a8ad7284d247513671628a4ac5917e104c/patches/react-dom-17+17.0.2.patch)

##  `renderToRenderStream` was removed

As you now have to `await` the `render` call, `renderToRenderStream` had no real value anymore.

Where previously, the second line of
```js
const renderStream = renderToRenderStream(<Component />, combinedOptions)
// this was optional in the past
const utils = await renderStream.renderResultPromise
```
could be omitted and could save you some code, now that second line would become required.

This now was not any shorter than calling `createRenderStream` and `renderStream.render` instead, and as both of these APIs now did the same thing in a different fashion, this would lead to confusion to no more benefit, so the API was removed.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -285,17 +261,45 @@ await expect(snapshotStream).toRerender()
...
+  using _disabledAct = disableActEnvironment()
+
+  // your test code here
+
+  // as soon as this scope is left, the environment will be cleaned up
+})
+```
+
+If you cannot use `using`, you can also manually call the returned `cleanup`
+function:
```
Comment (jerelmiller) [inline README.md:293]:
```suggestion
function. We recommend using `finally` to ensure the act environment is cleaned up if your test fails, otherwise it could leak between tests:
```

You might consider something like this addition somewhere in here so that others understand the importance of `finally` here, otherwise this might get ignored and cause other issues. Feel free to tweak my sentence and move it where it makes sense (perhaps a caution callout?), but I definitely think we should add a statement here explaining its importance.

File: src/disableActEnvironment.ts [hunk 2]
```diff
@@ -0,0 +1,171 @@
...
+   */
+  preventModification?: boolean
+
+  /**
+   * If `true`, will change the configuration of the testing library to
+   * prevent auto-wrapping e.g. `userEvent` calls in `act`.
+   *
+   * @default true
+   */
+  adjustTestingLibConfig?: boolean
```
Comment (jerelmiller) [inline src/disableActEnvironment.ts:22]:
```suggestion
  patchTestingLibConfig?: boolean
```

Perhaps just me, but I prefer "patch" instead of "adjust" since its a bit more technical term.

File: src/disableActEnvironment.ts [hunk 3]
```diff
@@ -0,0 +1,171 @@
...
+        writable: true,
+        configurable: true,
+      })
+    })
+    Object.defineProperty(
+      typedGlobal,
+      'IS_REACT_ACT_ENVIRONMENT',
+      getNewPropertyDescriptor(false, preventModification),
+    )
+  }
```
Comment (jerelmiller) [inline src/disableActEnvironment.ts:94]:
```suggestion
  const previous = typedGlobal.IS_REACT_ACT_ENVIRONMENT
  cleanupFns.push(() => {
    Object.defineProperty(typedGlobal, 'IS_REACT_ACT_ENVIRONMENT', {
      value: previous,
      writable: true,
      configurable: true,
    })
  })
  Object.defineProperty(
    typedGlobal,
    'IS_REACT_ACT_ENVIRONMENT',
    getNewPropertyDescriptor(false, preventModification),
  )
```

Are these brackets needed?

File: src/disableActEnvironment.ts [hunk 4]
```diff
@@ -0,0 +1,171 @@
+import {getConfig} from '@testing-library/dom'
+
+const dispose: typeof Symbol.dispose =
+  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
+  Symbol.dispose ?? Symbol.for('nodejs.dispose')
+
+export interface DisableActEnvironmentOptions {
```
Comment (jerelmiller) [inline src/disableActEnvironment.ts:7]: Should these options be mentioned in the README?

File: src/renderStream/createRenderStream.tsx [hunk 5]
```diff
@@ -1,11 +1,11 @@
 import * as React from 'rehackt'
 
-import {render as baseRender, RenderOptions} from '@testing-library/react'
+import {type RenderOptions} from '@testing-library/react/pure.js'
 import {Assertable, markAssertable} from '../assertable.js'
+import {renderWithoutAct, RenderWithoutActAsync} from '../renderWithoutAct.js'
```
Comment (jerelmiller) [inline src/renderStream/createRenderStream.tsx:5]:
```suggestion
import {renderWithoutAct, type RenderWithoutActAsync} from '../renderWithoutAct.js'
```

?

**Review by jerelmiller — APPROVED** [review]
Had a few suggestions, but nothing major. Great job investigating all of this!

File: src/renderStream/createRenderStream.tsx (hunk 5 above)
Comment (phryneas) [reply src/renderStream/createRenderStream.tsx:5]: Thx!

File: src/disableActEnvironment.ts (hunk 4 above)
Comment (phryneas) [reply src/disableActEnvironment.ts:7]: I would honestly not - there's a bunch of other, probably less used, options I don't mention in the README.

I see the README as the tool to quickly pick it up and JSDoc as API documentation here.

File: src/disableActEnvironment.ts (hunk 2 above)
Comment (phryneas) [reply src/disableActEnvironment.ts:22]: `patch` would imply to me "we do something creepy with internals that was not intended", while this really just sets two config options. 
I want to have this sound a little more harmless :)

File: src/disableActEnvironment.ts (hunk 3 above)
Comment (phryneas) [reply src/disableActEnvironment.ts:94]: Not 100% needed, but they are intentional here to restrict the scope of `previous` and group the core functionality (it also makes it collapsible in the editor as a bonus, which is something I actively use).

File: README.md [hunk 6]
```diff
@@ -285,17 +261,45 @@ await expect(snapshotStream).toRerender()
...
+> Keep in mind that if you use the `/pure` import, you have to call the
+> `cleanup` export manually after each test.
+
+## Usage side-by side with `@testing-library/react` or other tools that set `IS_REACT_ACT_ENVIRONMENT` or use `act`
+
+This library is written in a way if should not be used with `act`, and it will
+throw an error if `IS_REACT_ACT_ENVIRONMENT` is `true`.
 
-## A note on `act`.
+React Testing Library usually sets `IS_REACT_ACT_ENVIRONMENT` to `true`
```
[the remark this answers is not in the document]
Comment (phryneas) [reply README.md:273]: Yeah, technically it doesn't do that if you import from `/pure`, but then it also doesn't really work until you set it yourself. Good call.

File: README.md [hunk 7]
```diff
@@ -285,17 +261,45 @@ await expect(snapshotStream).toRerender()
...
 
-You might want to avoid using this library with `act`, as `act`
-[can end up batching multiple renders](https://github.com/facebook/react/issues/30031#issuecomment-2183951296)
-into one in a way that would not happen in a production application.
+To use this library side-by-side with React Testing Library, we ship the
+`disableActEnvironment` helper to undo these changes temporarily.
 
-While that is convenient in a normal test suite, it defeats the purpose of this
-library.
+It returns a `Disposable` and can be used together with the `using` keyword to
```
[the remark this answers is not in the document]
Comment (phryneas) [reply README.md]: put it on `"using" keyword`
